Add scripts for performance testing

kyleknap opened the pull request:
The following scripts were added:

* benchmark - Benchmark time, memory, cpu, network in/out for any script
* upload-file - Uploads a single file with s3transfer
* download-file - Downloads a single file with s3transfer
* benchmark-upload - Benchmarks the upload of a single file with s3transfer
* benchmark-download - Benchmarks the download of a single file with s3transfer

On a side note, I would much rather have my other PR's looked at than this one and have this one deprioritized. I figured that since I have these performance scripts that I have been using for a while now that I would push them up for other people to use because I have found them really handy.

cc @jamesls @JordonPhillips

jamesls commented on scripts/performance/benchmark:
```diff
@@ -0,0 +1,122 @@
+elif sys.platform == 'darwin':
+    INTERFACE = 'en0'
+else:
+    raise RuntimeError('Script cannot be run on %s' % sys.platform)
```
Just curious, this script could eventually be made to run on windows (psutils exposes everything we need right?)

JordonPhillips replied:
Yeah, the interface would be different though. It could actually be different if the user has a more complicated setup, such as if they're using en1 by default or if their interface has a different naming scheme.

kyleknap replied:
Yep this something we can look to do. I just have not done really any testing with these scripts on windows.

jamesls commented on scripts/performance/benchmark:
```diff
@@ -0,0 +1,122 @@
+            sent_rate = sent_delta / dt
+            recv_rate = recv_delta / dt
+
+            t = time.time()
```
Can't you just use `current_time` here?

kyleknap replied:
Yeah I will update.

jamesls commented on scripts/performance/benchmark:
```diff
@@ -0,0 +1,122 @@
+    parser.add_argument(
+        'script', help='The script to run for benchmarking')
+    parser.add_argument(
+        '--data-interval', default=0.1, type=float,
```
Did setting the interval to something higher (say 1 second) change any of the CPU percentages?  I wonder if collecting so frequently starts to introduce additional load on the system that could influence the numbers.

JordonPhillips replied:
The setting in the script I had been using was 1 second. A lower setting can be useful for quicker running commands, but yeah it does add a bit of overhead. I don't know that it would be that significant though. It'd be interesting to see the numbers.

kyleknap replied:
It was not really noticeable on the EC2 instance that I was running on, but I can do some quick testing if you want. The main issue is that what I have found CPU measurements have a relatively high standard deviation when compared to measurements of time and memory so it may take a bit more work to tell how much of an effect a change of the data interval will have on it.

jamesls replied:
I'd be inclined to stick with @JordonPhillips 1 second interval unless we really need .1 second granularity.  It can always be overriden with the `--data-interval` argument anyways.

kyleknap replied:
The CPU usage increased about 10 to 25 % for process lasting 15 to 90 seconds. So I would be fine with upping the interval.

jamesls commented on scripts/performance/benchmark:
```diff
@@ -0,0 +1,122 @@
+def run_script(args):
+    # Run the script in a separate process.
+    script_args = args.script.split(' ')
+    return subprocess.Popen(script_args)
```
Why not just use `shell=True`?  That way a user can use pipes and stdout/stdin right?  Seems like it's not an issue since you're recursing through children processes.

kyleknap replied:
I think it should work. Let me try it out to see if there is any issues with doing that.

jamesls commented:
Looks good, had some high level comments:

* I prefer having all the requirements in the top level directory, so you can see which requirements you may need to install.  What about moving this to a top level `requirements-perf.txt` or possibly a more general `requirements-dev.txt`?  The latter file could also include `requirements-test.txt`.

* Adding a `utils` module in the perf script requires the user needs to ensure that module is importable, so you either need to mess around with your python path or you need to run it from that directory.  Do you imagine we'd add a lot more utility functions?  It would be nice if all the scripts were self contained.

kyleknap commented:
I am fine with the general comments as well. Seems reasonable.

jamesls commented:
Flipping to incorporating feedback for now.

kyleknap commented:
@jamesls @JordonPhillips I think I addressed everything and I did some testing on the difference between 0.1 and 1 intervals for CPU usage (there was a noticeable change in CPU performance so I updated the default). Let me know what you think.

jamesls commented on scripts/performance/benchmark:
```diff
@@ -0,0 +1,124 @@
+            if p.status() == psutil.STATUS_ZOMBIE:
+                p.kill()
+                break
+            # Sleep for the data interval.
```
FWIW, these comments are distracting.  I think it's pretty clear that `time.sleep(data_internal)` is sleeping for the data_interval.

In general, I only prefer comments that explain non-obvious things about the code or *why* the code is written this particular way.

jamesls commented on scripts/performance/benchmark:
```diff
@@ -0,0 +1,124 @@
+Usage
+=====
+
+NOTE: Make sure you run ``pip install -r requirements.txt`` before running.
```
Isn't this `requirements-dev.txt` now?

jamesls commented on scripts/performance/benchmark:
```diff
@@ -0,0 +1,124 @@
+                current_time, memory_used, cpu_percent, sent_rate,
+                recv_rate]) + '\n')
+            f.flush()
+    sys.exit(0)
```
Sorry I missed this.  You generally shouldn't be `sys.exit()`'ing from functions.  Given the `if __name__...` block already has a `sys.exit`, you can just return the RC.

kyleknap replied:
Ahh right I did not see that either when I refactored this script from another script.

jamesls commented on scripts/performance/benchmark-download:
```diff
@@ -0,0 +1,171 @@
+        # Run the benchmark script now.
+        subprocess.check_call(benchmark_args)
+    finally:
+        # Remove the scratch directory.
```
Same thing as before.  These comments are distracting.  I think it's reasonable to assume a python person will know what `shutil.rmtree` does.

jamesls commented:
:shipit: Looks good.  I had an overarching stylistic comment that some of the code comments are distracting, so just something to watch out for going forward.

kyleknap commented:
I updated based on @jamesls comments.

@JordonPhillips let me know if you have any more feedback.

JordonPhillips commented:
⛵
